eip-0012.md

```diff
@@ -0,0 +1,324 @@
+
+### ergo.sign_data(addr: Address, message: String): Promise\<String>
+
+Errors: APIError, DataSignError
+
+Signs the generic data {data} encoded as a hex string using the private key associated with {addr} if possible. If the wallet does not own this address, an error is returned. This should ask for user consent and produce an error if rejected. The wallet should also implement a message signing protocol such as the proposed [EmIP-005](https://github.com/Emurgo/EmIPs/blob/5b00fce84f31eb763892186eb9c88739ec809333/specs/emip-005.md) in order to make this endpoint safer for use, and make it harder for the user to accidentally sign data such as transactions, blocks, etc. Returns the signed data as a hex-encoded string.
+
+TBD: Hash algorithm/format to use.
+
+### ergo.submit_tx(tx: SignedTx): Promise\<TxId>
+
+Errors: APIError, TxSendError
+
+Uses the wallet to send the given tx and returns the TxId for the dApp to track when/if it is included as its own responsibility. This is technically not mandatory but any wallet that is using such a bridge should have this functionality. Wallets can additionally choose to rate-limit this call, and can return an error if they did not or could not send the transaction.
+
+
+## Data Formats
+
+The initial design for most of these types are meant to be the same as the JSON types in the Ergo full node, which matches up as of now with the `sigma-rust` representations, but these definitions are the definitive types.
+
+### Address
+
+String as the standard base58 encoding of an address.
+
+### Box
+
+Transaction output box. Format is the same as `sigma-rust`:
+```
+type Box = {|
+  boxId: BoxId,
+  value: Value,
+  ergoTree: ErgoTree,
+  assets: TokenAmount[],
+  additionalRegisters: {| [string]: Constant |},
+  creationHeight: number,
+  transactionId: TxId,
+  index: number,
+|};
+```
+Additional registers have string ids "R4", "R5" to "R9".
+
+### BoxId
+
+Hex string of the box id.
+
+### Constant
+
+Uses `sigma-rust` rep - Hex-encoded bytes for `SigmaSerialize` of a constant.
+
+### ContextExtension
+
+Uses the full node JSON rep (id to hex-encoded Sigma-state value). Empty object is for P2PK.
+```
+type ContextExtension = {||} | {|
+    values: {| [string]: string |}
+|};
+```
+or the empty `{}` object for P2PK inputs.
+
+### ErgoTree
+
+Uses `sigma-rust` rep - Hex string representing serialized bytes of the tree using `SigmaSerialize`
+
+### DataInput
+
+Read-only input (e.g. oracle usage) that is not spendable. Uses `sigma-rust` JSON rep:
+```
+type DataInput = {|
+    boxId: BoxId,
+|};
+```
+
+### SignedInput
+
+A signed `UnsignedInput`. Uses full-node JSON rep:
+```
+type SignedInput = {|
+  boxId: BoxId,
+  spendingProof: ProverResult,
+|};
+```
+
+### UnsignedInput
+
+Uses full-node JSON rep:
+```
+type UnsignedInput = {|
+  boxId: BoxId,
+  extension: ContextExtension,
+|};
```

Comment:
Need to add the full content of `Box` to the input here that way the light wallet can properly show the user the content of the input (and we can validate the content is right by making sure it matches the box id passed in)

Reply:
I think the wallet should fetch the contents itself from its information instead. It is done this way with ErgoPay and is absolutely feasible.